### PR TITLE
(PUP-10819) error handling for 3x functions.

### PR DESCRIPTION
**Pull request by gimmyxd**
Updates error handling when a 3x function is deferred

### CODE REVIEW

File: lib/puppet/parser/scope.rb [hunk 1]
```diff
@@ -1041,7 +1041,11 @@ def method_missing(method, *args, &block)
     if respond_to? method
       send(method, *args)
     else
-      raise Puppet::DevError, _("Function %{name} not defined despite being loaded!") % { name: name }
+      if caller.any? { |line| line =~ /runtime3_support.rb/ }
+        raise Puppet::DevError, _("Cannot load function: '%{name}'. 3.x functions cannot be deffered.") % { name: name }
```
Comment (GabrielNagy) [inline lib/puppet/parser/scope.rb:1045]: typo: deferred

File: spec/integration/application/agent_spec.rb [hunk 2]
```diff
@@ -513,4 +513,28 @@
       th.kill # kill thread so we don't wait too much
     end
   end
+
+  context 'when deffering 3.x functions' do
```
Comment (GabrielNagy) [inline spec/integration/application/agent_spec.rb:517]: typo: deferred

**Review by GabrielNagy — APPROVED** [review]
looks good, just some typos

File: lib/puppet/parser/scope.rb [hunk 3]
```diff
@@ -1041,7 +1041,11 @@ def method_missing(method, *args, &block)
     if respond_to? method
       send(method, *args)
     else
-      raise Puppet::DevError, _("Function %{name} not defined despite being loaded!") % { name: name }
+      if caller.any? { |line| line =~ /runtime3_support.rb/ }
```
Comment (GabrielNagy) [inline lib/puppet/parser/scope.rb:1044]: `line.include?('runtime3_support.rb')` could be faster but I'm not sure

Comment (justinstoller) [reply lib/puppet/parser/scope.rb:1044]: Can we determine this without going through the entire stack trace? It's a very expensive operation in JRuby.

Comment (gimmyxd) [reply lib/puppet/parser/scope.rb:1044]: I'll check to see if this can be determined without searching in the stack trace.

**Comment by joshcooper** [comment]
Superseded in https://github.com/puppetlabs/puppet/pull/8477

**Pull request closed without merging**